source/Irrlicht/CGLTFMeshFileLoader.cpp

```diff
@@ -113,20 +115,134 @@ IAnimatedMesh* CGLTFMeshFileLoader::createMesh(io::IReadFile* file)
  * Documentation: https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#meshes
  * If material is undefined, then a default material MUST be used.
 */
-void CGLTFMeshFileLoader::loadPrimitives(
-		const MeshExtractor& parser,
-		SMesh* mesh)
+void CGLTFMeshFileLoader::MeshExtractor::loadMesh(
+		const std::size_t meshIdx,
+		CSkinnedMesh *mesh,
+		CSkinnedMesh::SJoint *parent) const
 {
-	for (std::size_t i = 0; i < parser.getMeshCount(); ++i) {
-		for (std::size_t j = 0; j < parser.getPrimitiveCount(i); ++j) {
-			auto indices = parser.getIndices(i, j);
-			auto vertices = parser.getVertices(i, j);
-
-			SMeshBuffer* meshbuf(new SMeshBuffer {});
-			meshbuf->append(vertices.data(), vertices.size(),
-				indices.data(), indices.size());
-			mesh->addMeshBuffer(meshbuf);
-			meshbuf->drop();
+	for (std::size_t j = 0; j < getPrimitiveCount(meshIdx); ++j) {
+		auto vertices = getVertices(meshIdx, j);
+		if (!vertices.has_value())
+			continue; // "When positions are not specified, client implementations SHOULD skip primitive’s rendering"
+
+		// Excludes the max value for consistency.
+		if (vertices->size() >= std::numeric_limits<u16>::max())
+			throw std::runtime_error("too many vertices");
+
+		// Apply the global transform along the parent chain.
+		for (auto &vertex : *vertices) {
+			parent->GlobalMatrix.transformVect(vertex.Pos);
+			// Apply scaling, rotation and translation (in that order) to the normal.
+			parent->GlobalMatrix.transformVect(vertex.Normal);
+			// Undo the translation, leaving us with scaling and rotation.
+			vertex.Normal -= parent->GlobalMatrix.getTranslation();
+			// Renormalize (length might have been affected by scaling).
+			vertex.Normal.normalize();
+		}
+		
+		auto maybeIndices = getIndices(meshIdx, j);
+		std::vector<u16> indices;
+		if (maybeIndices.has_value()) {
+			indices = std::move(maybeIndices.value());
+			for (u16 index : indices) {
+				if (index >= vertices->size())
+					throw std::runtime_error("index out of bounds");
+			}
+		} else {
+			// Non-indexed geometry
+			indices = std::vector<u16>(vertices->size());
+			for (std::size_t i = 0; i < vertices->size(); i += 3) {
+				// Reverse winding order per triangle
+				indices[i] = i + 2;
+				indices[i + 1] = i + 1;
+				indices[i + 2] = i;
+			}
+		}

```

Comment:
Extract into another function? Maybe `initializeIndices`? Not sure, `getIndices` might need a better name to help distinguish what they do.